Enable multiple outputs for haskell packages.

nc6 opened the pull request:
###### Motivation for this change
Enable multiple outputs for Haskell packages. This should address issues with references to GHC even when building static binaries.

Since creating this PR, I've spotted https://github.com/NixOS/nixpkgs/pull/27196, which is going in the same direction. I've lifted the things done in that PR into this. Here I'm trying to go somewhat further in having a split binary output for executables, since when we produce statically linked executables we don't want references to GHC.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

0xABAB commented:
For such a major feature, it would be nice to have a test.

vcunat commented:
What :-) The issue has been open for almost two years and now we have two implementations in a single day?

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -871,4 +871,9 @@ self: super: {
     postInstall = "rm $out/bin/mkReadme && rmdir $out/bin";
   });
 
+  # These claim to have executables, but don't
```
I don't believe that this is the right approach. Instead, that issue must be fixed in cabal2nix. The tool should simply not consider executables that have `buildable = false` when generating the appropriate Nix expression. I'll fix that ASAP and re-generate hackage-packages.nix. Anyhow, please don't include these changes in this PR.

domenkozar commented:
I suggest we first merge https://github.com/NixOS/nixpkgs/pull/27196 and then take this to the next level :)

nc6 commented:
Updated and rebased atop #27196 - I'll try to keep tracking that branch, unless @cleverca22 wants to incorporate these changes.

domenkozar commented:
This should reduce the closure-size of `happy` executable (which we don't build yet separately statically since there is no benefit from it now).

domenkozar commented:
Thinking about this change, we will probably need to pass `bin` output at least for `buildTools` (for example `happy`)

domenkozar commented:
@nc6 the initial PR has been merged, so this can be rebased (and I left one comment I think will have to be addressed before we start testing).

Fuuzetsu commented:
AFAIU we're not porting the other multiple outputs back to 17.09 so I'm unsure if this one should be in that milestone either seeing as it has that as dependency

globin commented:
Not back to 17.03, 17.09 is correct :)

nc6 committed:
Enable multiple outputs for Haskell packages.

nc6 committed:
Provide hscolour, happy, alex bin to buildInputs.

Use stdenv.lib.getBin instead of `.bin`

nc6 commented:
@domenkozar @peti PTAL. I have hopefully addressed the issue with .e.g. happy, alex and hscolour.

peti commented:
Test builds run at https://hydra.nixos.org/jobset/nixpkgs/pr-27209.

FRidh commented:
@peti could you disable these builds or significantly reduce the amount of shares it takes. As with the previous Haskell jobset it has 8 times the shares staging has thereby completely stalling staging.

peti commented:
@FRidh, that jobset has approx. 1% of the available shares on `hydra.nixos.org`. I don't see how that could possibly "completely stall" any other jobset. Also, please note that this jobset runs only once every couple of days -- namely when I manually trigger it --, whereas `staging` is evaluated several times per day. Therefore, I think it's appropriate for this jobset to have a slightly higher priority.

FRidh commented:
@peti 
Most branches are stall, so just consider shares.
- staging: 6000 (0.13% out of 4488072 shares)
- pr-27209: 50000 (1.11% out of 4488072 shares)

Staging is evaluated only once per day. Yes, you trigger this one just for now, but yesterday or the day before there was another one. For several reasons staging has been stuck for a month. Right now it seems to be in a good state again but its not advancing when other jobs are taking the build capacity.

peti commented:
I think it's perfectly fine to allocate more resources to this branch than to staging, because this branch will be building less frequently. If anything, the number of shared allocated to `staging` should be raised because 0.13% isn't a lot, really. But that's something you should probably discuss with Eelco.

peti commented:
~~~~
Run jailbreak-cabal to lift version restrictions on build inputs.
/nix/store/5h0dx5q0dzkm6d8kbdr5flplb2rgmsd1-stdenv/setup: line 73: /nix/store/9gbjrbj775dgjqjalbprgf4zqhk9qjrm-jailbreak-cabal-1.3.2/bin/jailbreak-cabal: No such file or directory
~~~~

peti commented on pkgs/development/compilers/ghc/7.10.2.nix:
```diff
@@ -34,7 +34,7 @@ stdenv.mkDerivation rec {
   };
 
-  buildInputs = [ ghc perl libxml2 libxslt docbook_xsl docbook_xml_dtd_45 docbook_xml_dtd_42 hscolour ];
+  buildInputs = [ ghc perl libxml2 libxslt docbook_xsl docbook_xml_dtd_45 docbook_xml_dtd_42 hscolour.bin ];
```
All those accesses to `.bin` should be replaced by use of `getBin`. These variables are function arguments, so we really don't know what kind of derivation will be passed there and cannot rely on the fact that the `bin` output actually exists. People might call this derivation passing `hscolour.bin` for all we know.

peti reviewed (changes requested):
Builds that need `jailbreak-cabal` are broken.

Fuuzetsu commented:
I updated the branch which fixes jailbreak-cabal as well as addresses problem with multiple outputs + shared executables. PATAL.

Fuuzetsu commented:
Hm, upon further inspection, we may want to either drop the default split to bin for all executables or also enable the split for other components otherwise we very easily get packages with separate bin with back-reference to things in main output and it's rejected by nix due to cycle. Either we should enable all parts as separate by default or none of them and let users pick and choose.

Fuuzetsu commented:
There is still a bit more work to be done here, will ping once it's ready for testing more.

Fuuzetsu committed:
Fix cabal2nix/stack2nix overrides

Use lib.makeBinPath.

Fuuzetsu commented:
I think this will do, I was able to build a large Haskell project with this branch with many dependencies, jailbreaking, cabal2nix, overrides &c.

Fuuzetsu commented:
Update: it appears I broke `ghcWithPackages` so still work pending.

nc6 committed:
Don't build separate bin output for alex.

It has files in /usr/share which reference the bin output.

nc6 committed:
Disable separate etc output for hscolour.

This is not needed, since `hscolour` has no output to `/etc`, and it
breaks certain old builds (see https://hydra.nixos.org/build/57925511)
because the older cabal used does not understand this flag.

nc6 committed:
Fix `hastache` build error.

`hastache` has no binary output but for `mkReadme`, so disable the
separate bin output and keep removing the `mkReadme`.

nc6 commented:
@peti I've fixed all the issues I could spot in the current hydra build - any chance you can retrigger?

peti commented:
I've told Hydra to evaluate the jobset, but so far it hasn't done it. Not sure what is going on.

vcunat commented:
Hydra's a bit overloaded, I guess.  Builds are queued now, but they might take days to finish.

nc6 commented:
Darn, I realise I failed to fix the hscolour issue properly. Hopefully all the other issues should be solved, though.

nc6 committed:
Add specific overrides for a number of failing packages.

Finish overriding bin output for broken packages.

Put stack bash completions in $bin.

nc6 commented:
Okay, I've disabled a separate bin output for the handful of packages which seem broken for it, and hopefully resolved the last issue with old versions of cabal. Luckily neither of these changes require a rebuild of the world, so should be fast to test.

Everything else failing seems to be on unrelated issues (e.g. missing dependencies).

peti commented:
Builds are running: https://hydra.nixos.org/eval/1385900

peti commented:
BTW, it might be a good idea to rebase this branch relative to the current `master` at some point. Right now, there'd be merge conflicts in the generic builder if we'd try to merge.

dezgeg commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -177,7 +197,15 @@ assert allPkgconfigDepends != [] -> pkgconfig != null;
 
-  outputs = if (args ? outputs) then args.outputs else ([ "out" ] ++ (optional enableSeparateDataOutput "data") ++ (optional enableSeparateDocOutput "doc"));
+  outputs = if (args ? outputs) then args.outputs else
+    ( [ "out" ]
```
Note that in rest of nixpkgs, if there is a `bin` output, it's always the first. This convention should probably be followed here.

That way people who write e.g. `${git-annex}/bin/git-annex` in their scripts won't have their scripts broken.

Fuuzetsu replied:
I didn't know the order mattered actually, interesting.

dezgeg commented on pkgs/top-level/all-packages.nix:
```diff
@@ -803,7 +803,7 @@ with pkgs;
       wrapProgram $out/bin/stack2nix \
-        --prefix PATH ":" "${git}/bin:${cabal2nix}/bin:${cabal-install}/bin:${stack}/bin"
+        --prefix PATH ":" "${git}/bin:${stdenv.lib.getBin cabal2nix}/bin:${stdenv.lib.getBin cabal-install}/bin:${stdenv.lib.getBin stack}/bin"
     '';
```
`lib.makeBinPath [ git cabal2nix cabal-install stack ]}`

dezgeg commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -489,4 +492,40 @@ self: super: builtins.intersectAttrs super {
+  stack = overrideCabal super.stack (drv: {
+    postInstall = ''
+      exe=$bin/bin/stack
+      mkdir -p $out/share/bash-completion/completions
```
Bash completions should go to `$bin` (so that they get installed to the user environment the same time as the binaries).

dezgeg commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -177,7 +197,15 @@ assert allPkgconfigDepends != [] -> pkgconfig != null;
+    ++ (optional enableSeparateDataOutput "data")
+    ++ (optional enableSeparateDocOutput "doc")
+    ++ (optional enableSeparateBinOutput "bin")
+    ++ (optional enableSeparateBinOutput "libexec")
```
That's a lot of outputs. Does e.g. the `libexec` one actually save any disk space in practice? E.g. if you have a case:

````
/nix/store/...-foo-bin/bin/foo
````
that refers to:
````
/nix/store/...-foo-libexec/libexec/foo-helper
````
then splitting the `libexec` output only helps in the case someone needs only `libexec` but not `bin` (which sounds quite rare to me).

vcunat replied:
Why not propagate inputs from .dev instead?

dezgeg commented on pkgs/development/compilers/ghc/8.0.1.nix:
```diff
@@ -33,7 +33,7 @@ stdenv.mkDerivation rec {
   ] ++ stdenv.lib.optional stdenv.isLinux ./ghc-no-madv-free.patch;
 
-  buildInputs = [ ghc perl hscolour sphinx];
+  buildInputs = [ ghc perl (stdenv.lib.getBin hscolour) sphinx];
```
Hmm. Typically the way output propagation done by the stdenv multiple-outputs code works is that bringing the `dev` output to scope also brings `bin`.

E.g. `curl` has:
````
outputs = [ "bin" "dev" "out" "man" "devdoc" ];
````
but still adding `curl` to `buildInputs` adds the `curl` binary to PATH.

With that said, of course only adding `bin` might reduce the build-time dependency closure significantly.

nc6 replied:
I've re-ordered the outputs now as per your other suggestion, so this may work. But I think it doesn't hurt to just ask for the `bin` output anyway, so will leave this here.

Fuuzetsu commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -177,7 +197,15 @@ assert allPkgconfigDepends != [] -> pkgconfig != null;
+    ++ (optional enableSeparateDataOutput "data")
+    ++ (optional enableSeparateDocOutput "doc")
+    ++ (optional enableSeparateBinOutput "bin")
+    ++ (optional enableSeparateBinOutput "libexec")
```
Main reason for splitting out `libexec` was to get it out of `out` such that any references to the binaries in `libexec` from other outputs would not create `someOutput -> libexec (out) -> someOutput (propagated inputs list)`.

On the other side, I don't see a real downside to not splitting out what we can.

nc6 commented:
OK, pushed a version rebased atop master and addressed comments by @dezgeg. Unfortunately this entails changes in `generic-builder.nix` and hence more rebuilds...

nc6 commented:
Pushed too soon, currently broken will fix...

nc6 committed:
Fix broken rebase.

nc6 commented:
Ready for testing again.

peti commented:
Builds are runninig ...

nc6 committed:
Fix git-annex.

nc6 commented:
@peti I've just pushed a change to fix the build of `git-annex`, which was (afaict) the only package remaining which fails due to this change (other failures are broken packages). As such, unless you have other comments, I believe this PR is in a good state to be merged!

zimbatm reviewed (approved):
looking good. any package still failing to build?

nc6 commented:
Rebased to address conflict.

nc6 commented:
@zimbatm Nothing failing to build (that I can see) due to problems introduced by this PR.

peti commented:
Um, did anyone test whether a `ghcWithPackages` environment constructed in this branch actually works? Because it certainly looks like it does not:

~~~
nix-shell -p "haskellPackages.ghcWithPackages (ps: [ps.mtl])" --run "ghc-pkg list mtl"
/nix/store/szz84j5k1dy3jdashis6ws28d8l8zxxb-ghc-8.0.2-with-packages/lib/ghc-8.0.2/package.conf.d
    (no packages)
~~~

MP2E commented:
Just finished a rebuild off of master since this was merged, ghcWithPackages is broken here as well. My xmonad config will no longer compile, among other things.

peti commented:
Since the entire Haskell infrastructure is essentially broken in `master`, I've pushed a revert in 89f5d52cf44baec14a7f542a605243777b16e7a5 so that we can fix these remaining issues properly without undue time pressure. There's `ghcWithPackages` to worry about, and then there's `ghcWithHoogle`, too, which I reckon will be broken as well.

domenkozar commented:
@nc6 do you plan to revisit this? It's a really good change, I'm willing to help.
